### PR TITLE
TEST/MAINT: Refactor unique axis verification.

### PR DESCRIPTION
**Pull request by ssanderson**
Break it into a standalone function that handles any pandas type.

Also uses `preprocess` to push the validation into a decorator, and refactors the tests to check all possible axis permutations.

### CODE REVIEW

File: tests/test_panel_daily_bar_reader.py [hunk 1]
```diff
@@ -13,34 +13,36 @@
...
+
 import pandas as pd
 
 from zipline.data.us_equity_pricing import PanelDailyBarReader
-from zipline.testing.fixtures import WithTradingEnvironment, ZiplineTestCase
+from zipline.testing import ExplodingObject
+from zipline.testing.fixtures import ZiplineTestCase
 
 
-class TestPanelDailyBarReader(WithTradingEnvironment, ZiplineTestCase):
```
Comment (ssanderson) [inline tests/test_panel_daily_bar_reader.py]: WithTradingEnvironment wasn't being used.

File: tests/test_panel_daily_bar_reader.py [hunk 2]
```diff
@@ -13,34 +13,36 @@
...
+class TestPanelDailyBarReader(ZiplineTestCase):
     def test_duplicate_values(self):
-        df = pd.DataFrame()
-        panel = pd.concat([pd.Panel({"X": df}), pd.Panel({"X": df})])
-
-        with self.assertRaises(ValueError) as e:
-            # panel's items has duplicates
-            PanelDailyBarReader(None, panel)
-
-        self.assertEqual("Duplicated items found: ['X']",
```
Comment (ssanderson) [inline tests/test_panel_daily_bar_reader.py]: I wanted to include the axis names in the message, which meant I had to update these entries, and it seemed simpler to just calculated the expected name from the transpose, which made it easer to just iterate over all possible transpositions.

File: tests/test_panel_daily_bar_reader.py [hunk 3]
```diff
@@ -13,34 +13,36 @@
...
-                         e.exception.message)
-
-        with self.assertRaises(ValueError) as e:
-            # panel's major axis has duplicates
-            PanelDailyBarReader(None, panel.swapaxes(0, 1))
-
-        self.assertEqual("Duplicated items found: ['X']",
-                         e.exception.message)
```
Comment (ssanderson) [inline tests/test_panel_daily_bar_reader.py]: `exception.message` is deprecated in py3

File: zipline/utils/input_validation.py [hunk 4]
```diff
@@ -25,6 +25,43 @@
 from zipline.utils.preprocess import preprocess
 
 
+def verify_indices_all_unique(obj):
```
Comment (ssanderson) [inline zipline/utils/input_validation.py]: I've wanted this logic elsewhere before, so it felt reasonable to put it in the `input_validation` utils.

File: zipline/data/us_equity_pricing.py [hunk 5]
```diff
@@ -710,17 +715,8 @@ class PanelDailyBarReader(DailyBarReader):
     first_trading_day : pd.Timestamp
         The first trading day in the dataset.
     """
+    @preprocess(panel=call(verify_indices_all_unique))
```
Comment (ssanderson) [inline zipline/data/us_equity_pricing.py]: I like `preprocess`.  It's a little magical, but it helps a lot for uncluttering functions, and I think it provides a nice declarative view of a function's preconditions.  I could see an argument for leaving the call inline here though.

**Comment by jbredeche** [comment]
LGTM.  thanks for the improvements, and even more so, the explanations!  Very clear and useful.

**Pull request merged**